frontend/src/metabase/containers/QuestionResultLoader.jsx

```diff
@@ -0,0 +1,97 @@
+import React from "react";
+import { defer } from "metabase/lib/promise";
+
+/*
+ * Question result loader
+ *
+ * Handle runninng, canceling, and reloading Question results
+ *
+ * @example
+ * <QuestionResultLoader question={question}>
+ * { ({ result, cancel, reload }) =>
+ *     <div>
+ *       { result && (<Visualization ... />) }
+ *
+ *       <a onClick={() => reload()}>Reload this please</a>
+ *       <a onClick={() => cancel()}>Changed my mind</a>
+ *     </div>
+ * }
+ * </QuestionResultLoader>
+ *
+ */
+export class QuestionResultLoader extends React.Component {
+  state = {
+    result: null,
+    cancel: null,
+  };
+
+  componentWillMount() {
+    this._loadResult(this.props.question);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // if the question is different, we need to do a fresh load
+    // TODO - is this the best way to check for new questions? any performance
+    // implications?
+    if (nextProps.question !== this.props.question) {
```

Comment:
It might be best to do a "deeper" comparison in case a calling component creates a new instead of `Question` but it hasn't actually changed.

A simple check could be:

```
if ((nextProps.question && nextProps.question.getUrl()) !== (this.props.question && this.props.question.getUrl))
```

I *think* that would work.

We also might want to eventually add a `noReload` prop or something to prevent automatic reloading, but if we think the default should be to reload then we can leave it out until we need it.